apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,43 +63,115 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
         /// <param name="consumer">The consumer to flush to, cannot be null.</param>
         /// <param name="bufferSize">The buffer size in bytes.</param>
         public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T> sizer, int bufferSize)
+            => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
         {
-            return new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _semaphore.Wait();
+                foreach (T item in items)
+                {
+                    _size += _sizer.GetSize(item);
+                    _items.Add(item);
+                    if (_size >= _bufferSize)
+                    {
+                        FlushesWithSemaphoreHeld();
+                    }
+                }
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
```

Comment:
(getting a bit frustrated with GitHub's only allowing me to comment on this at the original comment location, and not under your reply...also, the comment order is all out of whack for me)

Oh right, just because there is an attempt to continue on the original context, doesn't mean there will be one on which to continue. Got it.

Added to the list of potential analyzers.